bridges/snowbridge/primitives/inbound-queue/src/v2/converter.rs

```diff
@@ -77,92 +90,172 @@ impl<
-					},
-					EthereumAsset::ForeignTokenERC20 { token_id, value } => {
-						let asset_loc = ConvertAssetId::convert(&token_id)
-							.ok_or(ConvertMessageError::InvalidAsset)?;
-						let mut reanchored_asset_loc = asset_loc.clone();
-						reanchored_asset_loc
-							.reanchor(&GlobalAssetHubLocation::get(), &EthereumUniversalLocation::get())
-							.map_err(|_| ConvertMessageError::CannotReanchor)?;
-						let asset: Asset = (reanchored_asset_loc, *value).into();
-						assets.push(AssetTransfer::ReserveWithdraw(asset));
-					},
-				}
-			}
+		let topic = blake2_256(&("SnowbridgeInboundQueueV2", message.nonce).encode());
 
-			let topic = blake2_256(&("SnowbridgeInboundQueueV2", message.nonce).encode());
+		let prepared_message = PreparedMessage {
+			origin: message.origin.clone(),
+			claimer,
+			assets,
+			remote_xcm,
+			execution_fee: execution_fee_asset,
+			topic,
+		};
 
-			let prepared_message = PreparedMessage {
-				origin: message.origin.clone(),
-				claimer,
-				assets,
-				remote_xcm,
-				execution_fee: execution_fee_asset,
-				topic
-			};
+		Ok(prepared_message)
+	}
+}
 
-			Ok(prepared_message)
+/// Parse and (non-strictly) decode `raw` XCM bytes into a `Xcm<()>`.
+/// If decoding fails, return an empty `Xcm<()>`—thus allowing the message
+/// to proceed so assets can still be trapped on AH rather than the funds being locked on Ethereum
+/// but not accessible on AH.
+fn decode_raw_xcm(raw: &[u8]) -> Xcm<()> {
+	if let Ok(versioned_xcm) =
+		VersionedXcm::<()>::decode_with_depth_limit(MAX_XCM_DECODE_DEPTH, &mut raw.as_ref())
+	{
+		if let Ok(decoded_xcm) = versioned_xcm.try_into() {
+			return decoded_xcm;
 		}
+	}
+	// Decoding failed; allow an empty XCM so the message won't fail entirely.
+	Xcm::new()
+}
+
+/// Construct the `Xcm<()>` needed to create a new asset on Polkadot.
+/// Returns an error if the source network is not Ethereum and the target network is not Polkadot.
+fn create_asset_xcm<CreateAssetCall, CreateAssetDeposit, EthereumNetwork>(
+	token: &H160,
+	network: u8,
+	eth_value: u128,
+) -> Result<Xcm<()>, ConvertMessageError>
+where
+	CreateAssetCall: Get<CallIndex>,
+	CreateAssetDeposit: Get<u128>,
+	EthereumNetwork: Get<NetworkId>,
+{
+	let chain_id = match EthereumNetwork::get() {
+		NetworkId::Ethereum { chain_id } => chain_id,
+		_ => return Err(ConvertMessageError::InvalidNetwork),
+	};
+	let bridge_owner = EthereumLocationsConverterFor::<[u8; 32]>::from_chain_id(&chain_id);
+
+	let dot_asset = Location::new(1, Here);
+	let dot_fee: xcm::prelude::Asset = (dot_asset, CreateAssetDeposit::get()).into();
+
+	let eth_asset: xcm::prelude::Asset =
+		(Location::new(2, [GlobalConsensus(EthereumNetwork::get())]), eth_value).into();
+
+	let create_call_index: [u8; 2] = CreateAssetCall::get();
+
+	let asset_id = Location::new(
+		2,
+		[
+			GlobalConsensus(EthereumNetwork::get()),
+			AccountKey20 { network: None, key: (*token).into() },
+		],
+	);
+
+	match network {
+		0 => {
```

Comment:
This can be factored out into a separate method, such as `MessageToXcm::make_create_asset_for_polkadot`

Reply:
Addressed in https://github.com/vgeddes/polkadot-sdk/pull/13/commits/6a1f0cb2901d9496f27d3bbcaaaf0fba1f37016d.